HPC-9238: Upgrade to Node.js v20

Pl217 opened the pull request:
- Needs https://github.com/UN-OCHA/hpc-repo-tools/pull/31 merged+released first
- Needs https://github.com/UN-OCHA/hpc-api-core/pull/131 merged+released first

Pl217 committed:
HPC-9238: Specify minimal version of Node.js v20

We still keep v18 as a valid option, since Dependabot
runner image still uses it
https://github.com/dependabot/dependabot-core/blob/main/npm_and_yarn/Dockerfile

Pl217 committed:
HPC-9238: Upgrade to Node.js v20 in Docker images

Pl217 committed:
HPC-9238: Use Node.js v20 in GitHub CI

Pl217 committed:
Update GitHub actions to versions that use `node20` as default runtime

CodeQL Actions are yet to release v3, which will run on `node20`
https://github.com/github/codeql-action/issues/1959

Pl217 committed:
📝 Update Node requirements in docs

Pl217 committed:
⬆️ Upgrade `@types/node` to `v20.8.10`

Pl217 committed:
⬆️ Upgrade Prettier to v3

Pl217 committed:
⬆️ Upgrade `lint-staged` to `v15.0.2`

Pl217 committed:
⬆️ Upgrade `@unocha/hpc-repo-tools` to `v4.0.0`

Since ESLint and Typescript are peer
dependencies, we need to update them too

Pl217 committed:
Move ESLint config from JSON to JS

This is done to prepare for flat ESLint
config which will be rolled out in v9
https://eslint.org/blog/2023/10/flat-config-rollout-plans

Pl217 committed:
Auto-fix linting errors

This commit is output of running command:
`yarn eslint --quiet --fix .`

Pl217 committed:
Disable undesirable unicorn linting settings

Pl217 committed:
Reduce `@typescript-eslint/no-explicit-any` rule to warning level

Upon release of v6 of `@typescript-eslint/*` packages,
`no-explicit-any` started throwing errors, which is
a good default behavior, but not something we're ready to
embrace due to lots of `any` usages in generic code

https://github.com/typescript-eslint/typescript-eslint/discussions/7130

Pl217 committed:
Manually fix remaining linting errors

Pl217 committed:
⬆️ Upgrade `@unocha/hpc-api-core` to `v7.0.0`

Pl217 committed:
Ignore revision in which linting errors were auto-fixed

Pl217 committed:
Organize imports in TS files

This commit is the output of command:
`yarn prettier --write .`

Prettier plugin uses same feature that
VS Code's "Organize import" uses to
order the imports

Pl217 committed:
Ignore revision where mass re-ordering of imports happened

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 committed:
OPS-9737: Upgrade minimal Node.js v18 version

Node.js v18.18.2 has multiple security fixes,
so we don't want earlier versions

Pl217 committed:
Update `pg` to `v8.11.3`

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review
